tools/fp20compiler/main.cpp

```diff
@@ -34,27 +34,82 @@ nvparse_errors errors;
 int line_number;
 char * myin = 0;
 
+static char* find_at_line_start(char* haystack, char* cursor, const char* needle) {
+    cursor = strstr(cursor, needle);
+    while (cursor != NULL) {
+
+        // Accept if candidate is at beginning of file or line
+        if ((cursor == haystack) || (cursor[-1] == '\n')) {
+            return cursor;
+        }
+
+        cursor = strstr(cursor + 1, needle);
+    }
+    return NULL;
+}
+
 void translate(const char* s) {
     char* ns = strdup(s);
 
-    // printf("-- %s\b", s);
+    // Keep a cursor for line-counting
+    char* line_cursor = ns;
+    unsigned int line = 1;
 
-    char* ts_start = strstr(ns, "!!TS1.0");
-    if (ts_start) {
-        char* ts_end = strstr(ts_start, "// End of program");
-        assert(ts_end);
-        *ts_end = 0;
+    // Look for the first shader magic
+    char* shader_start = find_at_line_start(ns, ns, "!!");
 
-        ts10_init(ts_start);
-        ts10_parse();
-
-        *ts_end = '/';
+    // Warn the user if we couldn't find any shader at all
+    if (shader_start == NULL) {
+        fprintf(stderr, "no shaders found\n");
     }
 
-    char* rc_start = strstr(ns, "!!RC1.0");
-    if (rc_start) {
-        rc10_init(rc_start);
-        rc10_parse();
+    // Loop until we can't find a shader anymore
+    while (shader_start != NULL) {
+
+        // The next shader magic marks the end of the current shader; find it
+        char* line_end = strchr(shader_start, '\n');
+        char* next_shader_start = find_at_line_start(ns, line_end + 1, "!!");
+
+        // Create a copy of the current shader magic
+        if (line_end) { *line_end = '\0'; }
+        char* shader_magic = strdup(shader_start);
+        if (line_end) { *line_end = '\n'; }
+
+        // Count which line the current shader starts on
+        while (line_cursor < shader_start) {
+            line_cursor = strchr(line_cursor, '\n');
+            assert(line_cursor != NULL);
+            line_cursor++;
+            line++;
+        }
+
+        // Add information about shader section to output
+        printf("/* %s (line %u) */\n", shader_magic, line);
```

Comment:
nit: This can break if the input file has something like: `!!RC1.0 #*/ C-code /*`
I don't consider this a deal-breaker though, as it's unlikely to have such things in the first line.

I do not intend to fix / prevent this for now.

Reply:
This commit was removed, see https://github.com/XboxDev/nxdk/pull/115#discussion_r338232789.